consensus/src/liveness/leader_reputation.rs

```diff
@@ -202,18 +187,186 @@ impl ReputationHeuristic for ActiveInactiveHeuristic {
+                            let count = map.entry(failed_proposer).or_insert(0);
+                            *count += 1;
+                        }
+                    }
+                    Err(msg) => {
+                        warn!(
+                            "Failed proposer conversion from indices failed at epoch {}, round {}: {}",
+                            meta.epoch(),
+                            meta.round(),
+                            msg
+                        )
+                    }
+                }
+                map
+            },
+        )
+    }
+}
 
-        COMMITTED_PROPOSALS_IN_WINDOW.set(committed_proposals as i64);
-        COMMITTED_VOTES_IN_WINDOW.set(committed_votes as i64);
-        LEADER_REPUTATION_WINDOW_SIZE.set(history.len() as i64);
+/// If candidate appear in the history, it's assigned active_weight otherwise inactive weight.
+pub struct ActiveInactiveHeuristic {
+    author: Author,
+    active_weight: u64,
+    inactive_weight: u64,
+    aggregation: NewBlockEventAggregation,
+}
+
+impl ActiveInactiveHeuristic {
+    pub fn new(
+        author: Author,
+        active_weight: u64,
+        inactive_weight: u64,
+        window_size: usize,
+    ) -> Self {
+        Self {
+            author,
+            active_weight,
+            inactive_weight,
+            aggregation: NewBlockEventAggregation::new(window_size, window_size),
+        }
+    }
+}
+
+impl ReputationHeuristic for ActiveInactiveHeuristic {
+    fn get_weights(
+        &self,
+        epoch: u64,
+        candidates: &[Author],
+        history: &[NewBlockEvent],
+    ) -> Vec<u64> {
+        let votes = self.aggregation.count_votes(epoch, candidates, history);
+        let proposals = self.aggregation.count_proposals(epoch, history);
+
+        COMMITTED_PROPOSALS_IN_WINDOW.set(*proposals.get(&self.author).unwrap_or(&0) as i64);
+        COMMITTED_VOTES_IN_WINDOW.set(*votes.get(&self.author).unwrap_or(&0) as i64);
+        LEADER_REPUTATION_WINDOW_SIZE.set(proposals.values().sum::<i32>() as i64);
+
+        candidates
+            .iter()
+            .map(|author| {
+                if votes.contains_key(author) || proposals.contains_key(author) {
+                    self.active_weight
+                } else {
+                    self.inactive_weight
+                }
+            })
+            .collect()
+    }
+}
+
+/// Heuristic that looks at successful and failed proposals, as well as voting history,
+/// to define node reputation, used for leader selection.
+///
+/// We want to optimize leader selection to primarily maximize network's throughput,
+/// but we also, in combinatoin with staking rewards logic, need to be reasonably fair.
+///
+/// Logic is:
+///  * if proposer round failure rate within the proposer window is strictly above threshold, use failed_weight (default 1).
+///  * otherwise, if node had no proposal rounds and no successful votes, use inactive_weight (default 10).
```

Comment:
i'd imagine we give 10 to no proposal rounds but successful votes. if no successful votes, it probably should be on the lowest tier

Reply:
Right now we have 3 tiers:
- any activity, failures below threshold
- no activity, failures below threshold
- failures above threshold

That seems fine to me. If someone has voting activity, but no proposal activity - that means leader election didn't select him to be a leader, not that it was unable to propose (as failure would put him into last bucket irrespective)
So I don't think it would be fair to separate proposal activity from only voting activity. What do you think?

Reply:
@ikabiljo - In AIT1, we observed a bunch of validators successfully voting but were not able to ever create a successful proposal. In the current implementation, I would imagine we would have a flip-flip situation where slow validators are not given a chance to propose for a particular window period, but they again will be selected to vote after that window for some time. I am wondering if "no proposal" should be treated as harshly (or a bit less) as failed proposal because the leader election didn't select it as a leader as the validator failed to propose in past? 

I recall we also discussed weight-based techniques. where recent failures are punished harder than past failures. Is that something worth considering?